### PR TITLE
feat: Public dir support

### LINKED ISSUE #523: Add support for "public" directory
## Is your feature request related to a problem? Please describe.
From `nextjs` 9.0.6 and above the `static` directory is moved to `public/static`. Is not mandatory to move it, but the `static` directory will be deprecated in the future. [More information here.](https://github.com/zeit/next.js/blob/master/errors/static-dir-deprecated.md)

Moving the `static` folder to `public/static` throws the next error:
```
***/app/node_modules/next-i18next/dist/commonjs/config/create-config.js:54
        throw new Error("Default namespace not found at ".concat(defaultNSPath));
        ^

Error: Default namespace not found at ***/app/static/locales/es/common.json
    at _default (***/app/node_modules/next-i18next/dist/commonjs/config/create-config.js:54:15)
    at new NextI18Next (***/app/node_modules/next-i18next/dist/commonjs/index.js:52:46)
    at Object.<anonymous> (***/app/i18n.js:12:18)
    at Module._compile (internal/modules/cjs/loader.js:936:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:947:10)
    at Module.load (internal/modules/cjs/loader.js:790:32)
    at Function.Module._load (internal/modules/cjs/loader.js:703:12)
    at Module.require (internal/modules/cjs/loader.js:830:19)
    at require (internal/modules/cjs/helpers.js:68:18)
    at Object.<anonymous> (***/app/server.js:14:14)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

## Describe the solution you'd like
The path `***/app/static/locales/es/common.json` should be **`***/app/public/static/locales/es/common.json`**

## Describe alternatives you've considered
For an easy fix, we could keep as temp solution the locales at the static path.

## Additional context
...

### PR DESCRIPTION
**Pull request by nathanfriemel**
Took a pass at adding public folder support, looking for feedback:

- Changed default config localePath to public/static/locales
- Updated create config to look for combinedConfig localePath and if not found try to fall back to static/locales and print a deprecation console warning in the server config; client config attempts to remove `public/` from backend paths
- Updated tests

close #523

### CODE REVIEW

**Comment by nathanfriemel** [comment]
Didn't realize there was already a PR for this, feel free to close this one.

**Comment by isaachinman** [comment]
Hi @nathanfriemel, thanks for doing this work. I haven't heard from the author of #525 in awhile, so we can proceed with this PR if you're keen.

Can you please remove all the function return typings? I appreciate you want to fix TS warnings, but that should be handled in a separate PR as it just makes it harder to review this work. Once that's sorted I will review. Thanks!

**Comment by nathanfriemel** [comment]
Return types are gone.

File: examples/simple/package.json [hunk 1]
```diff
@@ -14,10 +14,6 @@
   "devDependencies": {},
   "dependencies": {
     "express": "^4.16.4",
-    "i18next": "^14.0.1",
-    "next": "^9.0.3",
-    "next-i18next": "link:../../",
-    "react": "link:../../node_modules/react",
-    "react-dom": "link:../../node_modules/react-dom"
+    "next-i18next": "link:../../"
```
Comment (isaachinman) [inline examples/simple/package.json]: What's going on here?

File: package.json [hunk 2]
```diff
@@ -68,8 +68,8 @@
     "@types/express": "^4.16.1",
     "@types/jest": "^24.0.16",
     "@types/jest-environment-puppeteer": "^4.0.0",
-    "@types/react": "^16.8.4",
-    "@types/react-dom": "^16.8.2",
+    "@types/react": "^16.9.11",
```
Comment (isaachinman) [inline package.json]: Are these @types changes necessary? If not, let's revert that please.

File: package.json [hunk 3]
```diff
@@ -115,6 +115,7 @@
   },
   "peerDependencies": {
     "next": ">= 7.0.0",
-    "react": ">= 16.8.0"
+    "react": ">= 16.8.0",
+    "react-dom": ">= 16.8.0"
```
Comment (isaachinman) [inline package.json]: Why was this added?

File: src/config/create-config.ts [hunk 4]
```diff
@@ -32,36 +33,47 @@ export default (userConfig) => {
...
     // Validate defaultNS
     // https://github.com/isaachinman/next-i18next/issues/358
     if (process.env.NODE_ENV !== 'production' && typeof combinedConfig.defaultNS === 'string') {
       const defaultNSPath = path.join(process.cwd(), `${localePath}/${defaultLanguage}/${combinedConfig.defaultNS}.${localeExtension}`)
       const defaultNSExists = fs.existsSync(defaultNSPath)
       if (!defaultNSExists) {
-        throw new Error(`Default namespace not found at ${defaultNSPath}`)
+        // if defaultNS doesn't exist, try to fall back to the deprecated static folder
+        // https://github.com/isaachinman/next-i18next/issues/523
+        if (fs.existsSync(path.join(process.cwd(), `${STATIC_LOCALE_PATH}/${defaultLanguage}/${combinedConfig.defaultNS}.${localeExtension}`))) {
```
Comment (isaachinman) [inline src/config/create-config.ts]: Can you move this logic out to a variable? Something like `const staticDirExists`.

File: src/config/create-config.ts [hunk 5]
```diff
@@ -32,36 +33,47 @@ export default (userConfig) => {
...
+          console.warn('Deprecation Warning - falling back to /static folder, deprecated in next@9.1.*')
```
Comment (isaachinman) [inline src/config/create-config.ts]: We've got our own logging util in `src/utils/console-message.ts`. Please use that where possible.

File: src/config/create-config.ts [hunk 6]
```diff
@@ -32,36 +33,47 @@ export default (userConfig) => {
...
     if (!combinedConfig.ns) {
       const getAllNamespaces = p => fs.readdirSync(p).map(file => file.replace(`.${localeExtension}`, ''))
-      combinedConfig.ns = getAllNamespaces(path.join(process.cwd(), `${localePath}/${defaultLanguage}`))
+      combinedConfig.ns = getAllNamespaces(path.join(process.cwd(), `${serverLocalePath}/${defaultLanguage}`))
     }
 
   } else {
 
+    // remove public/ prefix from client site config
+    const clientLocalePath = localePath.replace(/^public\//, '')
```
Comment (isaachinman) [inline src/config/create-config.ts]: I wonder how safe this is. We should probably check via something like `localePath.startsWith('/public/')`.

File: examples/simple/package.json (hunk 1 above)
Comment (nathanfriemel) [reply examples/simple/package.json]: Upgrading next to version 9.0.6+ was causing the tests to throw [this error](https://reactjs.org/docs/error-decoder.html/?invariant=321).  Seemed related to [this bug](https://github.com/zeit/next.js/issues/9022) in next, a work around was mentioned [here](https://github.com/zeit/next.js/issues/9022#issuecomment-552136320).  Wanted to test these changes against next 9.1, probably not important, could revert back to 9.0.3

Comment (isaachinman) [reply examples/simple/package.json]: Ah yeah, that's what I figured. I tried to upgrade myself, but ran into that bug.

I'm not understanding how this runs at all without `react` and `react-dom` as deps.

**Comment by nathanfriemel** [comment]
Changing the default to `public/static/locales` would probably simplify things.  I'm pretty sure the Next blog suggests just moving the static folder into the public folder so that you don't have to update import paths across your app.  I can make that change and probably undo a lot of the test changes in the process.

**Comment by isaachinman** [comment]
Sounds good. Thanks for being flexible!

File: src/config/create-config.ts (hunk 5 above)
Comment (nathanfriemel) [reply src/config/create-config.ts]: Will switch over.

File: src/config/create-config.ts (hunk 6 above)
Comment (nathanfriemel) [reply src/config/create-config.ts]: Should be a no-op if the pattern doesn't match, but couldn't hurt to check startsWith first.

File: src/config/create-config.ts (hunk 4 above)
Comment (nathanfriemel) [reply src/config/create-config.ts]: Will do.

File: package.json (hunk 2 above)
Comment (nathanfriemel) [reply package.json]: Reverting

File: package.json (hunk 3 above)
Comment (nathanfriemel) [reply package.json]: Reverting

**Comment by isaachinman** [comment]
@nathanfriemel Just let me know when this is ready for review again.

**Comment by nathanfriemel** [comment]
@isaachinman I think it is ready to look at again.

File: __tests__/config/test-helpers.ts [hunk 7]
```diff
@@ -15,24 +15,24 @@ const userConfig = {
   defaultNS: 'universal',
   fallbackLng: 'it',
   otherLanguages: ['fr', 'it'],
-  localePath: 'static/translations',
+  localePath: 'public/translations',
```
Comment (isaachinman) [inline __tests__/config/test-helpers.ts]: Any reason why you didn't make this `public/static/translations`?

Comment (nathanfriemel) [reply __tests__/config/test-helpers.ts]: Just missed it, will update it and the tests.

**Comment by nathanfriemel** [comment]
@isaachinman README updated and updated test-helpers localePath.

File: __tests__/config/test-helpers.ts [hunk 8]
```diff
@@ -15,24 +15,24 @@ const userConfig = {
...
-  localePath: 'static/translations',
+  localePath: 'public/static/translations',
   localeStructure: '{{ns}}/{{lng}}',
   localeSubpaths: localeSubpathVariations.FOREIGN,
 }
 
 const userConfigClientSide = {
   ...userConfig,
   backend: {
-    loadPath: '/static/translations/{{ns}}/{{lng}}.json',
```
Comment (isaachinman) [inline __tests__/config/test-helpers.ts]: Why was this change made?

File: __tests__/config/test-helpers.ts [hunk 9]
```diff
@@ -15,24 +15,24 @@ const userConfig = {
...
+    addPath: '/translations/{{ns}}/{{lng}}.missing.json',
   },
 }
 
 const userConfigServerSide = {
   ...userConfig,
   backend: {
-    loadPath: '/home/user/static/translations/{{ns}}/{{lng}}.json',
-    addPath: '/home/user/static/translations/{{ns}}/{{lng}}.missing.json',
+    loadPath: '/home/user/public/translations/{{ns}}/{{lng}}.json',
```
Comment (isaachinman) [inline __tests__/config/test-helpers.ts]: Again wondering why this isn't `public/static`?

File: src/utils/console-message.ts [hunk 10]
```diff
@@ -31,9 +31,9 @@ function logMessage(messageType, message) {
  * @param {String} message
  * @param {Object} options
  */
-export default function createConsoleLog(messageType, message) {
+export default function createConsoleLog(messageType, message, config = this.config) {
```
Comment (isaachinman) [inline src/utils/console-message.ts]: Why was this change made? I suppose it's because we're trying to reference `this.config` before it exists?

File: yarn.lock [hunk 11]
```diff
@@ -1900,7 +1900,7 @@ async-sema@3.0.0:
   resolved "https://registry.yarnpkg.com/async-sema/-/async-sema-3.0.0.tgz#9e22d6783f0ab66a1cf330e21a905e39b3b3a975"
```
Comment (isaachinman) [inline yarn.lock]: Why is there a diff on the lockfile?

File: __tests__/config/test-helpers.ts (hunk 9 above)
Comment (nathanfriemel) [reply __tests__/config/test-helpers.ts]: It should be public/static, also the create-config.test should be changed to public/static on lines 115-116.

Looks like process.env.NODE_ENV is always 'production' in the create-config.test even with `NODE_ENV=test jest` so the test never runs the `Validate defaultNS` block and serverLocalePath is just whatever the test-helpers userConfig.localePath is set to.  I'm not sure why NODE_ENV is production atm and interested to see if you see the same thing.

File: src/utils/console-message.ts (hunk 10 above)
Comment (nathanfriemel) [reply src/utils/console-message.ts]: Using consoleMessage in create-config was throwing `Cannot read property 'config' of undefined`

File: yarn.lock (hunk 11 above)
Comment (nathanfriemel) [reply yarn.lock]: Now that all the package.json changes are reverted there shouldn't be any diff, I'll revert it to master.

Comment (isaachinman) [reply yarn.lock]: Cool, thanks!

File: __tests__/config/test-helpers.ts (hunk 9 above)
Comment (isaachinman) [reply __tests__/config/test-helpers.ts]: Yeah, it looks like it's due to [this](https://github.com/isaachinman/next-i18next/blob/e48f7e9558c341d3f1fd476e56acd5f0433cad52/__tests__/config/create-config.test.ts#L249) hacky setup.

We should instead capture the original value in a var, and reassign before the end of the test:

```jsx
const originalNodeEnv = process.env.NODE_ENV
process.env.NODE_ENV = 'production'

...perform test

process.env.NODE_ENV = originalNodeEnv
```

Modifying globals in tests is almost always problematic!

Comment (isaachinman) [reply __tests__/config/test-helpers.ts]: cc @capellini

**Comment by isaachinman** [comment]
@nathanfriemel Need to resolve a couple conflicts unfortunately. But this PR is my top priority at the moment, so I won't merge anything except any potential urgent bug fixes until we get it in.

File: __tests__/config/test-helpers.ts (hunk 9 above)
Comment (capellini) [reply __tests__/config/test-helpers.ts]: Hey @isaachinman - do you need something from me?

Comment (isaachinman) [reply __tests__/config/test-helpers.ts]: Just tagged you to notify you that we'll most likely need to update these tests. If @nathanfriemel is happy to do the work, I don't think we need anything!

Comment (capellini) [reply __tests__/config/test-helpers.ts]: Ok.  I'm here to assist, if needed. Just let me know.

Comment (isaachinman) [reply __tests__/config/test-helpers.ts]: Thanks!

Comment (nathanfriemel) [reply __tests__/config/test-helpers.ts]: Resolved the conflicts, want me to go ahead and change static to public/static?

**Comment by isaachinman** [comment]
@nathanfriemel Let me know when this ready for re-review. Thanks!

File: __tests__/config/test-helpers.ts (hunk 9 above)
Comment (nathanfriemel) [reply __tests__/config/test-helpers.ts]: Went ahead and changed to public/static

**Comment by nathanfriemel** [comment]
@isaachinman Resolved conflicts, ready to look at.

File: src/config/create-config.ts [hunk 12]
```diff
@@ -33,36 +34,53 @@ export const createConfig = (userConfig) => {
...
       const defaultNSExists = fs.existsSync(defaultNSPath)
       if (!defaultNSExists) {
-        throw new Error(`Default namespace not found at ${defaultNSPath}`)
+        // if defaultNS doesn't exist, try to fall back to the deprecated static folder
+        // https://github.com/isaachinman/next-i18next/issues/523
+        const staticDirPath = path.join(process.cwd(), STATIC_LOCALE_PATH, defaultFile)
+        const staticDirExists = fs.existsSync(staticDirPath)
+        if (staticDirExists) {
+          consoleMessage('warn', 'Falling back to /static folder, deprecated in next@9.1.*', combinedConfig)
+          serverLocalePath = STATIC_LOCALE_PATH
```
Comment (isaachinman) [inline src/config/create-config.ts]: I am just now realising that this fallback will only take place in non-production modes. Wouldn't that effectively break all currently-deployed apps where people rely on the default value of `localePath` being the `static` dir?

Comment (nathanfriemel) [reply src/config/create-config.ts]: Yeah you're right, will look at this again in the morning.

Comment (nathanfriemel) [reply src/config/create-config.ts]: Sorry for the break, holidays and work and what not.  What was the original reason for the `process.env.NODE_ENV !== 'production'` here?  Was it to prevent the error throw in production?  I looked at the ticket in the comment and the PR for it and just want to figure out the original intent before changing anything more.

Comment (isaachinman) [reply src/config/create-config.ts]: No problem. Yes that's correct - this `if` block is/was solely for non-production validation of the default namespace.

Comment (nathanfriemel) [reply src/config/create-config.ts]: Moved the production check to just before the error throw, hopefully that sorts things.

**Comment by isaachinman** [comment]
Thank you @nathanfriemel for all your work on this. I merged the PR and cleaned up a few things, and have released under v3.0.0. There should be no breaking changes, just wanted to be on the safe side.
